File: frontend/src/lib/types/vendorStatementRecon.ts
// Types for the Vendor Statement Reconciliation surface. Mirrors the JSON
// returned by the `/api/vendor-statements` endpoints (the Pydantic contract in
// `backend/app/schemas/vendor_statement_recon.py`). Money fields arrive as
// numbers (or null); date/time fields are ISO strings (or null).

import type { BadgeTone } from '$lib/components/ui/Badge.svelte';

// --- Run status -----------------------------------------------------------

export type ReconStatus = 'open' | 'resolved';

export const RECON_STATUSES: ReconStatus[] = ['open', 'resolved'];

// StatusBadge-style label map (Title Case) for the run status pill.
export const RECON_STATUS_LABELS: Record<ReconStatus, string> = {
	open: 'Open',
	resolved: 'Resolved'
};

// Badge tone per status, so the list page and the modal can't tint the same
// status two different shades — which is exactly what they did (.12 alpha on
// the list, .15 in the modal, off two different ambers).
export const RECON_STATUS_TONES: Record<ReconStatus, BadgeTone> = {
	// An open run is work still owed to the ledger, not an error — amber.
	open: 'warning',
	resolved: 'success'
};

// --- Intake source --------------------------------------------------------

// How the run's statement lines got here. `manual` = typed into the lines
// editor; `csv` = a deterministic parse of an uploaded CSV; `pdf` = MACHINE-READ
// off an uploaded document by the org's extraction adapter. The distinction is
// not cosmetic: a reviewer clearing a `pdf` run's lines is clearing a model's
// reading of a document, which is why the provenance block exists.
export type ReconSourceFormat = 'manual' | 'csv' | 'pdf';

// Data-value map, English by the established convention (see the status /
// classification maps above).
export const RECON_SOURCE_FORMAT_LABELS: Record<ReconSourceFormat, string> = {
	manual: 'Entered by hand',
	csv: 'CSV upload',
	pdf: 'PDF (machine-read)'
};

// --- Line classification --------------------------------------------------

export type ReconClassification =
	| 'matched'
	| 'amount_mismatch'
	| 'missing_on_our_side'
	| 'missing_on_their_side';

export const RECON_CLASSIFICATION_LABELS: Record<ReconClassification, string> = {
	matched: 'Matched',
	amount_mismatch: 'Amount mismatch',
	missing_on_our_side: 'Missing (our side)',
	missing_on_their_side: 'Missing (their side)'
};

// --- Line resolution ------------------------------------------------------

export type ReconResolutionStatus = 'unresolved' | 'resolved' | 'ignored';

export const RECON_RESOLUTION_LABELS: Record<ReconResolutionStatus, string> = {
	unresolved: 'Unresolved',
	resolved: 'Resolved',
	ignored: 'Ignored'
};

// --- Response shapes ------------------------------------------------------

export interface ReconLine {
	id: string;
	classification: ReconClassification;
	resolution_status: ReconResolutionStatus;
	statement_invoice_number: string | null;
	statement_date: string | null;
	statement_amount: number | null;
	statement_status: string | null;
	matched_invoice_id: string | null;
	matched_invoice_number: string | null;
	ledger_amount: number | null;
	amount_difference: number | null;
	match_method: string | null;
	resolution_note: string | null;
	resolved_at: string | null;
}

export interface ReconSummary {
	line_count: number;
	matched_count: number;
	amount_mismatch_count: number;
	missing_our_side_count: number;
	missing_their_side_count: number;
	statement_total: number | null;
	ledger_total: number | null;
}

// Provenance for a run whose lines were MACHINE-READ off a PDF. Present only on
// the PDF intake path; `null` for CSV / pasted-lines runs, which have neither a
// provider nor a confidence to report.
export interface StatementExtractionMeta {
	method: string;
	provider: string;
	// 0..1 as the adapter reported it — render through `formatExtractionConfidence`.
	confidence: number;
	// How many open items the reader actually accepted off the document. NOT the
	// run's line count: the run also carries `missing_on_their_side` rows built
	// from our own ledger.
	line_count: number;
	// How many rows the reader recognised as an open item but REFUSED to book
	// because it could not resolve them (a second money column, a second
	// reference column). Deliberately not a count of every skipped line —
	// headers, totals and page furniture are skipped silently. Optional because
	// runs created before the reader counted them carry no such key.
	skipped_ambiguous?: number;
}

export interface Reconciliation {
	id: string;
	vendor_id: string | null;
	vendor_name: string | null;
	statement_date: string;
	statement_reference: string | null;
	currency: string;
	source_format: string;
	file_key: string | null;
	// Whether the supplier's own document was archived alongside the run —
	// fetched by run id via `GET /{id}/file`, never by key.
	has_source_file: boolean;
	extraction: StatementExtractionMeta | null;
	status: ReconStatus;
	notes: string | null;
	summary: ReconSummary;
	created_at: string;
	updated_at: string | null;
	// Lines are included on the detail response only (the list omits them).
	lines?: ReconLine[] | null;
}

export interface ReconciliationListResponse {
	items: Reconciliation[];
	total: number;
	page: number;
	page_size: number;
}

/**
 * Whole-set KPI rollup from `GET /api/vendor-statements/summary`, over the SAME
 * vendor_id / status filters the list ran with.
 *
 * The page's `openCount` filtered the LOADED page and `totalDiscrepancies`
 * reduced the per-run discrepancy counts over it — both contradicting the
 * whole-set `total`. `open_discrepancies` is that reduce, whole-set.
 */
export interface ReconciliationSummary {
	total: number;
	by_status: Record<string, number>;
	open_discrepancies: number;
}

// --- Create payloads ------------------------------------------------------

// One supplier-statement line for the manual / pasted-lines intake path.
export interface StatementLineInput {
	invoice_number?: string | null;
	invoice_date?: string | null;
	amount?: number | null;
	status?: string | null;
}

// POST /api/vendor-statements body (JSON, pasted-lines path).
export interface ReconciliationCreate {
	vendor_id: string;
	statement_date: string;
	statement_reference?: string | null;
	currency?: string;
	notes?: string | null;
	lines: StatementLineInput[];
}

// --- Line resolve ---------------------------------------------------------

export interface LineResolveRequest {
	resolution_status: ReconResolutionStatus;
	resolution_note?: string | null;
}

// --- Close-readiness ------------------------------------------------------

export interface CloseReadinessVendor {
	vendor_id: string | null;
	vendor_name: string | null;
	reconciliation_id: string;
	statement_date: string;
	currency: string;
	unreconciled_amount: number;
	missing_our_side_count: number;
	amount_mismatch_count: number;
}

export interface CloseReadinessResponse {
	materiality_threshold: number;
	blocking_vendors: CloseReadinessVendor[];
	is_close_ready: boolean;
}

// --- Pure display helpers -------------------------------------------------

/**
 * Render an adapter's 0..1 confidence as a whole percentage.
 *
 * Defensive because the figure crosses a network boundary: a missing / NaN /
 * non-numeric value renders the placeholder rather than `NaN%`, and an
 * out-of-range value is clamped rather than shown as `140%` — a reviewer
 * weighing a machine's reading must never be handed a nonsense number.
 * Deliberately NOT locale-formatted: it is a provenance qualifier, not a
 * measured figure, and it must not be confused with money.
 */
export function formatExtractionConfidence(
	confidence: number | null | undefined,
	placeholder = '—'
): string {
	if (typeof confidence !== 'number' || !Number.isFinite(confidence)) return placeholder;
	const clamped = Math.min(1, Math.max(0, confidence));
	return `${Math.round(clamped * 100)}%`;
}

/** Were this run's lines read off a document by an extraction adapter? */
export function isMachineRead(recon: Pick<Reconciliation, 'extraction'>): boolean {
	return recon.extraction !== null && recon.extraction !== undefined;
}

/**
 * How many supplier rows the reader saw but refused to book, or `0`.
 *
 * Crosses a network boundary and is absent on runs created before the reader
 * counted its skips, so anything that isn't a finite non-negative integer reads
 * as zero — the panel must never announce `NaN rows were skipped`, and must not
 * announce a skip warning on a run that simply predates the field. Rounded
 * rather than trusted: a count is not a measured figure.
 */
export function ambiguousSkipCount(
	extraction: StatementExtractionMeta | null | undefined
): number {
	const raw = extraction?.skipped_ambiguous;
	if (typeof raw !== 'number' || !Number.isFinite(raw) || raw <= 0) return 0;
	return Math.round(raw);
}

/**
 * Filename to save the archived supplier document under.
 *
 * Composed from the run's own metadata rather than parsed out of `file_key` —
 * the key is an internal storage detail (the API returns the `has_source_file`
 * flag precisely so a client needn't reach into it), and `source_format` already
 * says which document shape was uploaded.
 */
export function sourceStatementFilename(
	recon: Pick<Reconciliation, 'source_format' | 'statement_date' | 'vendor_name'>
): string {
	const ext = recon.source_format === 'pdf' ? 'pdf' : 'csv';
	const vendor = (recon.vendor_name ?? 'vendor')
		.toLowerCase()
		.replace(/[^a-z0-9]+/g, '-')
		.replace(/^-+|-+$/g, '');
	return `statement-${vendor || 'vendor'}-${recon.statement_date}.${ext}`;
}

File: backend/app/schemas/vendor_statement_recon.py
"""Pydantic schemas for vendor statement reconciliation.

Shared contract for the ``/api/vendor-statements`` router and the frontend
``/vendor-statements`` route. Money fields use the ``MoneyAmount`` annotations
(Decimal in Python, JSON number on the wire); IDs are strings on the wire,
parsed to UUID in the router. See
``backend/docs/vendor-statement-reconciliation.md``.
"""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field

from app.schemas.money import MoneyAmount, OptionalMoneyAmount


class StatementLineInput(BaseModel):
    """One supplier-statement line, for the manual / pasted-lines intake path.

    ``invoice_number`` is the join key the engine matches on; ``amount`` is the
    open balance the supplier claims. Both date and status are free-form on a
    supplier statement, so they're optional.
    """

    invoice_number: str | None = Field(default=None, max_length=100)
    invoice_date: date | None = None
    # Digits match `vendor_statement_recon_lines.statement_amount` Numeric(18, 2).
    amount: MoneyAmount | None = Field(default=None, ge=0, max_digits=18, decimal_places=2)
    status: str | None = Field(default=None, max_length=40)


class ReconciliationCreate(BaseModel):
    """Create a reconciliation run from a pasted/normalised list of lines."""

    vendor_id: str
    statement_date: date
    statement_reference: str | None = Field(default=None, max_length=120)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    notes: str | None = Field(default=None, max_length=500)
    lines: list[StatementLineInput] = Field(default_factory=list)


class ReconLineResponse(BaseModel):
    id: str
    classification: str
    resolution_status: str
    statement_invoice_number: str | None
    statement_date: str | None
    statement_amount: OptionalMoneyAmount = None
    statement_status: str | None
    matched_invoice_id: str | None
    matched_invoice_number: str | None = None
    ledger_amount: OptionalMoneyAmount = None
    amount_difference: OptionalMoneyAmount = None
    match_method: str | None
    resolution_note: str | None
    resolved_at: str | None


class ReconciliationSummary(BaseModel):
    """The denormalised outcome rollup carried on every run response."""

    line_count: int
    matched_count: int
    amount_mismatch_count: int
    missing_our_side_count: int
    missing_their_side_count: int
    statement_total: OptionalMoneyAmount = None
    ledger_total: OptionalMoneyAmount = None


class StatementExtractionMeta(BaseModel):
    """Provenance for a run whose lines were MACHINE-READ off a PDF.

    A reviewer clearing these lines is clearing a model's reading of a
    document, not a supplier's typed CSV — so which provider read it and how
    confident it was travels with the run. ``None`` on the CSV / pasted-lines
    paths, which have neither.
    """

    method: str
    provider: str
    confidence: float
    line_count: int
    # Rows the reader saw as an open item but refused to book because it could
    # not resolve them (two money columns, a second reference column). NOT a
    # count of every skipped line — headers, totals and page furniture are
    # skipped silently, and counting those would report noise. Defaulted so a
    # run persisted before this field existed still deserialises.
    skipped_ambiguous: int = 0


class ReconciliationResponse(BaseModel):
    id: str
    vendor_id: str | None
    vendor_name: str | None
    statement_date: str
    statement_reference: str | None
    currency: str
    source_format: str
    file_key: str | None
    # The archived supplier document is fetched by run id, never by key — the
    # flag is what a client needs; the key is an internal detail.
    has_source_file: bool = False
    extraction: StatementExtractionMeta | None = None
    status: str
    notes: str | None
    summary: ReconciliationSummary
    created_at: str
    updated_at: str | None
    # Lines are included on the detail response only (list omits them).
    lines: list[ReconLineResponse] | None = None


class ReconciliationListResponse(BaseModel):
    items: list[ReconciliationResponse]
    total: int
    page: int
    page_size: int


class ReconciliationSummaryResponse(BaseModel):
    """Whole-set KPI rollup for the reconciliation list — counterpart of
    ``GET /api/expenses/summary``. Takes the SAME ``vendor_id`` / ``status``
    filters as the list (via ``_recon_list_filters``). The page's ``openCount``
    filtered the LOADED page and ``totalDiscrepancies`` reduced the per-run
    discrepancy counts over it — both contradicting the whole-set ``total``.

    ``open_discrepancies`` sums ``amount_mismatch + missing_our_side +
    missing_their_side`` across the filtered set — the exact figure the page's
    ``discrepancyCount`` reduce produced, just whole-set."""

    total: int
    by_status: dict[str, int]
    open_discrepancies: int


class LineResolveRequest(BaseModel):
    """Resolve or ignore one reconciliation line. ``resolved`` = the difference
    has been actioned (e.g. the missing invoice was created); ``ignored`` = a
    known, accepted discrepancy."""

    resolution_status: str = Field(..., pattern="^(resolved|ignored|unresolved)$")
    resolution_note: str | None = Field(default=None, max_length=500)


class CloseReadinessVendor(BaseModel):
    """One vendor flagged as not-close-ready: its most recent run still carries
    a material unreconciled balance."""

    vendor_id: str | None
    vendor_name: str | None
    reconciliation_id: str
    statement_date: str
    currency: str
    unreconciled_amount: MoneyAmount
    missing_our_side_count: int
    amount_mismatch_count: int


class CloseReadinessResponse(BaseModel):
    materiality_threshold: MoneyAmount
    blocking_vendors: list[CloseReadinessVendor]
    is_close_ready: bool
